src/components/join-flow/username-step.jsx

```diff
@@ -92,12 +94,37 @@ class UsernameStep extends React.Component {
             }
         );
     }
+    // memoize remote requests for weak password check
+    validatePasswordRemotelyWithCache (password) {
+        if (typeof this.passwordRemoteCache[password] === 'object') {
+            return Promise.resolve(this.passwordRemoteCache[password]);
+        }
+        // password is not in our cache
+        return validate.validatePasswordRemotely(password).then(
+            remoteResult => {
+                // cache result, if it successfully heard back from server
+                if (remoteResult.requestSucceeded) {
+                    this.passwordRemoteCache[password] = remoteResult;
+                }
+                return remoteResult;
+            }
+        );
+    }
     validatePasswordIfPresent (password, username) {
         if (!password) return null; // skip validation if password is blank; null indicates valid
-        const localResult = validate.validatePassword(password, username);
-        if (localResult.valid) return null;
-        return this.props.intl.formatMessage({id: localResult.errMsgId});
-    }
+        // if password is not blank, run both local and remote validations
+        const localResult = validate.validatePasswordLocally(password, username);
+        return this.validatePasswordRemotelyWithCache(password).then(
+            remoteResult => {
+                if (localResult.valid === false) { // defer to local check first
+                    return this.props.intl.formatMessage({id: localResult.errMsgId});
```

Comment:
Since this deferral happens (by definition) without regard to the remote result, is there a reason to make the remote validation request in the first place, if the local result guaranteed an overall failure anyway?

An early return seems like it should be fine, i.e. by moving this `if (localResult.valid === false)` check above the call to `validatePasswordRemotelyWithCache`.

(This suggested change would mean the surrounding function `validatePasswordIfPresent` would sometimes return sync and sometimes async, so you could do `return Promise.resolve(formatMessage(...))`, but Formik [supports sync/async validation anyway](https://formik.org/docs/guides/validation) - this PR already takes advantage of that by not making any changes to areas where `validatePasswordIfPresent` is referenced.)

Reply:
Agreed -- if localResult.valid is false, we should not call validatePasswordRemotelyWithCache